### ISSUE TITLE
[ENH] Add optional margin for CDW-CE

### ISSUE THREAD

**Issue by conradbkay** [ayrna/dlordinal#111]
### Describe the feature or idea you want to propose

Paper by some of the original CDW-CE authors:
[https://arxiv.org/abs/2412.01246](https://arxiv.org/abs/2412.01246)

The formula doesn't make sense to me assuming `log(1 − max(1, yˆi + m))` is always undefined

### Describe your proposed solution

I'm guessing they mean something like 
```py
if self.margin is not None:
    probs = F.softmax(y_pred, dim=1)
    probs = torch.clamp(probs + self.margin, max=1 - 1e-8)
    log_term = torch.log(1 - probs)
    
    loss = log_term * weights
```

**Comment by victormvy** [comment]
Hi! The authors of CDW-CE with margin confirmed by email that there is a typo in the equation in the paper.

Where it is written as:

`log(1 - max(1, y_i + m))`

it should read:

`log(1 - min(1 - y_i + m, 1))`.

I have implemented this correction in the linked PR, including an optimisation that computes the loss directly from logits instead of first applying softmax and using probabilities (as already done in the non-margin version).
